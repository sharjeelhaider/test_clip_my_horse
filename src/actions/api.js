import {asyncActionNames, buildAsyncActions} from './utils/asyncUtils.js';
import {encode} from 'base-64';
// Constants
export const LOAD_HORSES = asyncActionNames('LOAD_HORSES');

// Actions
export const receiveHorses = buildAsyncActions(LOAD_HORSES);

// Async actions
export const fetchData = (
  username,
  password,
  endpoint,
  fetchAction,
) => async dispatch => {
  let auth = {Authorization: 'Basic ' + encode(username + ':' + password)};
  const requestOptions = {
    method: 'GET',
    headers: auth,
  };
  try {
    dispatch(fetchAction.request());
    const response = await fetch(endpoint, requestOptions);
    if (handleErrors(response)) {
      const responseJson = await response.json();
      dispatch(fetchAction.success(responseJson));
    }
  } catch (error) {
    dispatch(fetchAction.failure(error));
  }
};

function handleErrors(response) {
  if (!response.ok) {
    throw Error('Authentication error');
    //Api always return null for statusText
    // throw Error(response.statusText);
  }
  return response;
}
