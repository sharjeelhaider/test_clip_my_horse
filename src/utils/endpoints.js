export const horses =
  'https://xooz0jxyu6.execute-api.us-east-1.amazonaws.com/v1/horses';
