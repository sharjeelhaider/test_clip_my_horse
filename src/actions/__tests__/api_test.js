import mockStore from 'redux-mock-store';

import {fetchData} from '../api';
import {asyncActionNames, buildAsyncActions} from '../utils/asyncUtils';

const store = mockStore();
const TEST_API = asyncActionNames('TEST_API');
const receiveTestData = buildAsyncActions(TEST_API);

beforeEach(() => {
  store.clearActions();
});

it('should handle TEST_API_SUCCESS action', async () => {
  const response =
    '{"items": [{"id": "df579a48-a7ca-4afc-89d0-caabcfac87d7"}]}';
  fetch.mockResponseSuccess(response);
  await store.dispatch(
    fetchData('foo@gmail.com', 'bar', '/test', receiveTestData),
  );
  expect(store.getActions()).toMatchSnapshot();
});

it('should handle TEST_API_FAILURE action', async () => {
  const mockError = new Error('Look ma! I am an error.');
  fetch.mockResponseFailure(mockError);
  await store.dispatch(
    fetchData('foo@gmail.com', 'bar', '/test', receiveTestData),
  );
  expect(store.getActions()).toMatchSnapshot();
});
