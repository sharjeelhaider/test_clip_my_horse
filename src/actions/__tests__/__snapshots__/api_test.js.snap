// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle TEST_API_FAILURE action 1`] = `
Array [
  Object {
    "type": "TEST_API_REQUEST",
  },
  Object {
    "error": [Error: Look ma! I am an error.],
    "type": "TEST_API_FAILURE",
  },
]
`;

exports[`should handle TEST_API_SUCCESS action 1`] = `
Array [
  Object {
    "type": "TEST_API_REQUEST",
  },
  Object {
    "error": [Error: Authentication error],
    "type": "TEST_API_FAILURE",
  },
]
`;
